conda/exceptions.py

```diff
@@ -444,34 +445,77 @@ def __init__(self, channel):
 
 class UnavailableInvalidChannel(ChannelError):
 
-    def __init__(self, channel, error_code):
+    def __init__(self, channel, status_code, response=None):
         from .models.channel import Channel
         from .common.url import join_url, maybe_unquote
+
+        # parse channel
         channel = Channel(channel)
         channel_name = channel.name
         channel_url = maybe_unquote(channel.base_url)
-        message = dals("""
-        The channel is not accessible or is invalid.
-          channel name: %(channel_name)s
-          channel url: %(channel_url)s
-          error code: %(error_code)d
 
-        You will need to adjust your conda configuration to proceed.
-        Use `conda config --show channels` to view your configuration's current state,
-        and use `conda config --show-sources` to view config file locations.
-        """)
+        # define hardcoded/default reason/message
+        reason = getattr(response, "reason", None)
+        message = dals(
+            """
+            The channel is not accessible or is invalid.
+
+            You will need to adjust your conda configuration to proceed.
+            Use `conda config --show channels` to view your configuration's current state,
+            and use `conda config --show-sources` to view config file locations.
+            """
+        )
+        if channel.scheme == "file":
+            url = join_url(channel.location, channel.name)
+            message += dedent(
+                f"""
+                As of conda 4.3, a valid channel must contain a `noarch/repodata.json` and
+                associated `noarch/repodata.json.bz2` file, even if `noarch/repodata.json` is
+                empty. Use `conda index {url}`, or create `noarch/repodata.json`
+                and associated `noarch/repodata.json.bz2`.
+                """
+            )
+
+        # if response includes a valid json body we prefer the reason/message defined there
+        try:
+            body = response.json()
+        except (AttributeError, JSONDecodeError):
+            body = {}
+        else:
+            reason = body.get("reason", None) or reason
+            message = body.get("message", None) or message
+
+        # standardize arguments
+        status_code = status_code or "000"
+        reason = reason or "UNAVAILABLE OR INVALID"
+        if isinstance(reason, str):
+            reason = reason.upper()
+
+        # compose complete message
+        message = (
+            dals(
+                f"""
+                HTTP {status_code} {reason} for channel {channel_name} <{channel_url}>
+
+                """
+            )
+            # since message may include newlines don't include in f-string/dals above
+            + message
+        )
+
+        from .auxlib.logz import stringify
```

Comment:
Would it be possible to move this to the top of the file? It's a real pet-peeve of mine to see these import statements here, especially when it's not necessary (e.g. because of circular reference issues)

Reply:
Absolutely, this was result of a copy-paste that I didn't rethink.